allennlp/service/predictors/predictor.py

```diff
@@ -21,16 +19,28 @@ class Predictor(Registrable):
     a ``Predictor`` is a thin wrapper around an AllenNLP model that handles JSON -> JSON predictions
     that can be used for serving models through the web API or making predictions in bulk.
     """
-    def __init__(self,
-                 model: Model,
-                 tokenizer: Tokenizer,
-                 token_indexers: Dict[str, TokenIndexer]) -> None:
-        self.model = model
-        self.tokenizer = tokenizer
-        self.token_indexers = token_indexers
+    def __init__(self, model: Model, dataset_reader: DatasetReader) -> None:
+        self._model = model
+        self._dataset_reader = dataset_reader
 
     def predict_json(self, inputs: JsonDict) -> JsonDict:
-        raise NotImplementedError()
+        instance = self._json_to_instance(inputs)
+        outputs = self._model.forward_on_instance(instance)
+        return self._predictions_to_json(outputs)
+
+    def _json_to_instance(self, json: JsonDict) -> Instance:
```

Comment:
The reason was for inference that needs to happen after `model.forward`.  Seems like there are a few options here:

1. We do what I did here, which is to put the call to inference / decoding logic in the `Predictor`, in a method that happens after the call to `model.forward_on_instance`.  It really should be an overridable method, because not every model needs to do this inference / decoding, and how that logic works depends on the model and the predictor.
2. We put the inference / decoding logic into `model.forward_on_instance`.  Either this means that models that do some kind of decoding need to put that decoding into `model.forward()` itself, or they need to override `model.forward_on_instance` to do the decoding there.  Because this probably means we're adding it to `model.forward()`, you probably want some kind of flag that says whether to do inference, because you probably don't want to bother at training time, but you might for the validation set.
3. We add another method on `Model` called `decode`, or something, that takes the output of `model.forward_on_instance` (or `model.forward`, but one of these has pytorch variables and the other has numpy - doing the conversion inside of `decode` wouldn't be hard...), and adds a decoded field to the dictionary.  The default implementation in the base class would just return an un-modified dictionary.  Then we can just add a call to `model.decode()` in `Predictor.predict_json()`, and remove the `_predictions_to_json` stuff.

I like option 3, I think.  It gives a cleaner separation between the jobs of `model.forward()` (which computes potentials / probabilities and a loss) and a decoding step (which does some kind of max inference / marginalization / beam search / whatever, given the potentials), and it makes it easy to have the logic all live inside of `Model`, where it belongs, without making training slow by putting unnecessary decoding into `forward`.

What do you think?

Reply:
Yep 3 sounds good. We should also consider whether we want this function to be able to take a list of instances and run forward on all of them as a batch or something as well.

Reply:
Yeah, that's a good point.  I'll start with just a single instance, and we can worry about expanding it to a batched case later when someone needs it.

Reply:
+1 to having all the logic live inside the `Model`